### ISSUE TITLE
yubikey pin/admin pin thoughts

### ISSUE THREAD

**Issue by Valodim** [open-keychain/open-keychain#1393]
warning, wall of text incoming :hurtrealbad: 

so, one important part of yubikey workflows is how we handle pin and admin pin. dominik and I discussed this, but we'd like to hear more thoughts on this if anyone cares (ping @josecastillo and @jas4711)

so, a couple of general thoughts first:
- usability, usability, usability. this is the prime directive.
- we want to provide a truly complete solution for yubikeys at some point, the user should not have to do any part of the workflow with gpg
- use of default pins (123456 or 12345678 admin) is unacceptable
- 6 digits is the minimum length for pins, 8 for admin pins ([click](https://github.com/Yubico/ykneo-openpgp/blob/master/applet/src/openpgpcard/OpenPGPApplet.java#L78))
- no pin is required for setup if we do a factory reset
- (for the record) even though they are called "pins", technically any password can be used. we will stick to numeric pins though

following these thoughts, here's the workflow we came up with:
- in the yubikey creation wizard, the yubikey will first be factory reset. this follows a "might as well" logic, since we don't care about special scenarios like only changing specific subkeys
- we will then set the pin and admin pin to generated ones with 6 respectively 10 digits. (it makes no sense to ask the user to input their own here)
- the first 6 digits of the admin pin are the same as the regular pin(!!)
- the user will be asked to write down the admin pin, and in the next dialog asked to input both pins for verification

the idea here is that from a security point of view, the admin pin has strictly more privileges than the regular pin, and can be used to brute force the regular pin. if an attacker finds the written down admin pin he gains no useful information by also knowing the regular pin at that point. on the other hand, the user will have written down the regular pin implicitly, which helps remembering it later, and allows the user to look it up if they forget instead of being required to change it.

a thought along the way was that 4 digit pins would actually suffice since 6 digits are already hard to remember and three attempts rule out brute forcing, but this is not something we can do due to the required minimum of 6 digits.

**Comment by jas4711** [comment]
This seems OK to me.  Thanks for discussing before implementing!   Anyone gaining access to the admin pin will be able to gain access anyway, since they can just enter an invalid pin three times and then reset the pin.  A subtlety is that someone who gains access to the admin code cannot in general gain access without that being detected unless they also gain access to the PIN.  With your scheme, that property is modified.  It is a subtle security weakening, but I think it is minor.

Re 4 vs 6 digits.  I'm not sure where the min requirement on pin length comes from, I have only ever used 6-digits pin's with my various OpenPGP cards.  I suspect the openpgp card spec says something about it?  If there is no required limit, I think we can consider lowering this to 4 digits since I don't see any problem ALLOWING option that for users that understand what they are doing.  This won't affect a lot of deployed keys though.

/Simon

**Comment by Valodim** [comment]
thanks for your input!

re admin suffix. you are right about that subtle security weakening, and I thought about that as well. the precise difference is that with the pin, the attacker can access the openpgp operations on contact, with the admin pin it takes them a while (few minutes at most) to do so, since they first need to brute-force the pin. this is an acceptable trade-off in my eyes, and users who don't find this acceptable can still change the pin manually.

re 4 vs 6 digits. the 6 character minimum is indeed specified in the openpgp card specs, section 4.2, so we will have to live with it.

we will also probably offer a "remember this pin" option, essentially modifying the "use default pin (123456)" preference to "default pin to use, if any". this mitigates the drive-by vulnerability while conserving maximum convenience, for a minimum attack scenario who would have to steal both the yubikey and the phone.

**Comment by jas4711** [comment]
Re admin: the attacker doesn't need to brute force anything, does it?  if you know the admin pin, you just take the first 6 digits and you have the user PIN.

Yeah, I think your app should (after a reset) modify the default PIN to something random and then offer to remember that directly.

Reset'ing the applet is a bit evil, but I understand why you want that from a usability and simplicity point of view.  Just remember to REALLY clearly say that the user is losing all their keys on that NEO.  Preferrably you should try to read out any existing information from the card and display that, so the user doesn't user the wrong key by mistake.

Hm. I just remembered that reseting the applet takes a considerably amount of time, and likely burns quite some power in the process.  I'm not sure how well that will work over NFC, really.  Please test that too.

**Comment by josecastillo** [comment]
@jas4711, I have noticed something in some of my tests with the ykneo-openpgp applet running on a smart card (I haven't repeated this on an actual NEO): a power loss in between the TERMINATE and ACTIVATE commands seems to put the applet in an unrecoverable state. I can't replicate it in the simulator, but this would be something to test if apps like OpenKeychain are going to attempt to reset the card over the NFC interface, where power loss during a long-running operation is more likely. Here's [a sample session using gpg-connect-agent](https://gist.github.com/josecastillo/666abfb3eddb657926e1); if you can replicate this, it might be an argument against resetting the applet over NFC.

**Comment by jas4711** [comment]
@klali - what do we want to recommend here?  I think reseting the applet over NFC is likely to be problematic, but I cannot quote chapter and verse as to why it is a bad idea.

**Comment by Valodim** [comment]
we're still planing this out and didn't test anything yet. wouldn't have expected this to be a very complex operation, isn't it just memsetting a bunch of memory to zero? if there is any chance of bricking yubikeys though it is obviously not an option.

usability-wise it would be a really neat option though, since "setup from scratch" and "import existing" are really the only relevant scenarios we need to cover. people who want to have a special subkey setup will use gpg anyways, regardless of what we do, so we will not sacrifice simplicity to support whatever weird scenarios the pgp bloggers come up with that month

re admin: I meant for the case where the admin pin is independent from the regular pin, that required brute forcing is the difference in security.

**Comment by jas4711** [comment]
re admin: okay, got it.  i agree.  brute force should be difficult because I think it is supposed to lock you out after a couple of attempts.

re reset: For various reasons, I think the reset operation actually is more complicated than memsetting.  I'll prod @klali to give his thoughts tomorrow.

**Comment by klali** [comment]
So the main problem with the reset over NFC (that I can think of right now) is that the activate ins (0x44) does a memory reclaim after initializing objects. That takes a couple of seconds, and if it's interrupted it will continue next time the javacard is activated. I think the only way to see how that actually behaves and if the usability is ok is to test it.

@josecastillo it looks like that error is caused by gnupg becoming confused when it wants to do a status-fetch before actually sending instruction 44 (the apdu sent is 00 CA 00 4F 00 with the 69 85 reply making it sad). Looking at the openpgpcard 2.0 spec the ykneo-openpgp applet isn't conforming exactly to the behaviour after terminate (select should return 62 85) on the other hand, changing that doesn't seem to make a difference to gnupg, same behaviour in not recognising the applet after the select, so never sending the instructed apdu.

**Comment by klali** [comment]
and regarding pin lengths, the spec says:
"The OpenPGP application uses two local passwords for user verification, called PW (PW1 with 6 characters minimum and PW3 with 8 characters minimum)"

**Comment by Valodim** [comment]
I have another question: I thought I read that some resellers preset pins and send them to the customer together with the yubikey, similar to SIM card pins. I can't find the source where I read that so I'm not sure, can you confirm or deny this?

**Comment by dschuermann** [comment]
I decided upon the following:
* PIN can now be chosen, but we check for insecure ones like 123456 etc before the user can proceed
* Admin PIN is auto generated and the user is prompted to write it down and store it safely.
This resembles how PIN and PUK usage of sim cards. Nowadays people can choose their own PIN.

In addition:
* I removed the "Use default PIN" option, it is confusing and dangerous. Users are encouraged to set a PIN and not use the default one
* I implemented and tested reset over NFC and it works pretty good. I also took away the YubiKey between the two reactivate commands, still it can not be bricked with this. In contrast to GnuPG I only check if both commands succeed not after every command (see https://github.com/open-keychain/open-keychain/commit/709a91422e0ecaf0099d257288ea80305f7002b3#diff-577b99776553b6267a7fa8a4f34e56c7R718).
